### PR TITLE
Feat/model backed ui

### PR DESCRIPTION
**Pull request by kevcunnane**
This is an initial PR for a new model-driven UI where extensions can provide definitions of the components & how they're laid out using Containers. 
#1140, #1141, #1142, #1143 and #1144 are all tracking additional work needed to improve the initial implementation and fix some issues with the implementation.

Features:
- Supports defining a FlexContainer that maps to a flexbox-based layout. 
- Supports creating a card component, which is a key-value pair based control that will lay out simple information to a user. Eventually this will have an optional set of actions associated with it.
- Has a sample project which shows how to use the API and was used for verification

### CODE REVIEW

File: samples/sqlservices/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# README
+## This is the README for your extension "sqlservices"
```
Comment (kburtram) [inline samples/sqlservices/README.md]: this readme could probably be more personalized

File: samples/sqlservices/src/apiWrapper.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export class ApiWrapper {
+	// Data APIs
+
+	public registerWebviewProvider(widgetId: string, handler: (webview: data.DashboardWebview) => void): void {
+		return data.dashboard.registerWebviewProvider(widgetId, handler);
+	}
+
+
+	public registerControlHostProvider(widgetId: string, handler: (webview: data.DashboardWebview) => void): void {
+		return data.dashboard.registerWebviewProvider(widgetId, handler);
```
Comment (kburtram) [inline samples/sqlservices/src/apiWrapper.ts]: should this call registerControlHostProvider?

File: samples/sqlservices/src/controllers/mainController.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+		sqlops.tasks.registerTask('sqlservices.clickTask', (profile) => {
+			vscode.window.showInformationMessage(`Clicked from profile ${profile.serverName}.${profile.databaseName}`);
+		});
+
+		return Promise.resolve(true);
+	}
+
+	private registerSqlServicesModelView(): void {
+		sqlops.dashboard.registerModelViewProvider('sqlservices', async (view) => {
```
Comment (kburtram) [inline samples/sqlservices/src/controllers/mainController.ts]: probably should be separate from dashboard, though we can make those refactors later.

File: samples/sqlservices/tasks/buildtasks.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    return gulp.src(srcFiles)
+        .pipe(srcmap.init())
+        .pipe(tsProject())
+        .on('error', function() {
+            if (process.env.BUILDMACHINE) {
+                done('Extension Tests failed to build. See Above.');
+                process.exit(1);
+            }
+        })
+        //.pipe(nls.rewriteLocalizeCalls())
```
Comment (kburtram) [inline samples/sqlservices/tasks/buildtasks.js]: dead code

File: src/sql/parts/dashboard/contents/mvvm/modelComponentWrapper.component.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	ngAfterViewInit() {
+		this.updateTheme(this._bootstrap.themeService.getColorTheme());
+		if (this.componentHost) {
+			this.loadComponent();
+		}
+		this._changeref.detectChanges();
+		this.layout();
+	}
+
+	// public refresh(): void {
```
Comment (kburtram) [inline src/sql/parts/dashboard/contents/mvvm/modelComponentWrapper.component.ts]: dead code

File: src/sql/parts/dashboard/contents/mvvm/modelComponentWrapper.component.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+			this._changeref.detectChanges();
+		} catch (e) {
+			error('Error rendering component: {0}', e);
+			return;
+		}
+		let el = <HTMLElement>componentRef.location.nativeElement;
+
+		// set widget styles to conform to its box
+		el.style.overflow = 'hidden';
+		// el.style.flex = '1 1 auto';
```
Comment (kburtram) [inline src/sql/parts/dashboard/contents/mvvm/modelComponentWrapper.component.ts]: dead code

File: samples/sqlservices/typings/sqlops.proposed.d.ts [hunk 7]
```diff
@@ -0,0 +1,203 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+// This is the place for API experiments and proposal.
+
+import * as core from 'sqlops';
```
Comment (kburtram) [inline samples/sqlservices/typings/sqlops.proposed.d.ts]: do we really need to duplicate this file in the sample?  won't it be difficult to keep up to date?

File: src/sql/sqlops.proposed.d.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+		taskId: string;
+	}
+
+	export interface CardProperties  {
+		label: string;
+		value?: string;
+		actions?: ActionDescriptor[];
+
+	}
+	export interface CardComponent extends Component {
```
Comment (llali) [inline src/sql/sqlops.proposed.d.ts]: Please add in label, value getters and setters here

File: samples/sqlservices/src/apiWrapper.ts (hunk 2 above)
Comment (kevcunnane) [reply samples/sqlservices/src/apiWrapper.ts]: Removing this as it's not really needed / used for this sample

File: samples/sqlservices/src/controllers/mainController.ts (hunk 3 above)
Comment (kevcunnane) [reply samples/sqlservices/src/controllers/mainController.ts]: Agreed, will take later

File: samples/sqlservices/typings/sqlops.proposed.d.ts (hunk 7 above)
Comment (kevcunnane) [reply samples/sqlservices/typings/sqlops.proposed.d.ts]: I can remove and use as relative. Good suggestion

**Pull request merged**